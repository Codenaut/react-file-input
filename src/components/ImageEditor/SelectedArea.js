import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import '../../styles/ImageEditor.scss';

class SelectedArea extends PureComponent {
  constructor(props) {
    super(props);

    this.selectedArea = null;

    this.startMove = this.startMove.bind(this);

    this.resizeTopLeft = this.resizeTopLeft.bind(this);
    this.resizeTop = this.resizeTop.bind(this);
    this.resizeTopRight = this.resizeTopRight.bind(this);
    this.resizeRight = this.resizeRight.bind(this);
    this.resizeBottomRight = this.resizeBottomRight.bind(this);
    this.resizeBottom = this.resizeBottom.bind(this);
    this.resizeBottomLeft = this.resizeBottomLeft.bind(this);
    this.resizeLeft = this.resizeLeft.bind(this);
  }

  resizeTopLeft(event) {
    const { startResize } = this.props;

    startResize(event, true, true, false);
  }

  resizeTop(event) {
    const { startResize } = this.props;

    startResize(event, false, true, true, true);
  }

  resizeTopRight(event) {
    const { startResize } = this.props;

    startResize(event, false, true);
  }

  resizeRight(event) {
    const { startResize } = this.props;

    startResize(event, false, false, true);
  }

  resizeBottomRight(event) {
    const { startResize } = this.props;

    startResize(event, false, false);
  }

  resizeBottom(event) {
    const { startResize } = this.props;

    startResize(event, false, false, true, true);
  }

  resizeBottomLeft(event) {
    const { startResize } = this.props;

    startResize(event, true, false);
  }

  resizeLeft(event) {
    const { startResize } = this.props;

    startResize(event, true, false, true);
  }

  startMove(event) {
    if (event.target === this.selectedArea) {
      const { startMove } = this.props;

      startMove(event);
    }
  }

  render() {
    const { style } = this.props;

    return (
      <div
        className="brainhub-image-editor__selected-area"
        ref={ref => {
          this.selectedArea = ref;
        }}
        style={style}
        onMouseDown={this.startMove} >
        <div
          className="brainhub-image-editor__resize-point brainhub-image-editor__resize-point--tl"
          onMouseDown={this.resizeTopLeft}
        />
        <div
          className="brainhub-image-editor__resize-point brainhub-image-editor__resize-point--t"
          onMouseDown={this.resizeTop}
        />
        <div
          className="brainhub-image-editor__resize-point brainhub-image-editor__resize-point--tr"
          onMouseDown={this.resizeTopRight}
        />
        <div
          className="brainhub-image-editor__resize-point brainhub-image-editor__resize-point--r"
          onMouseDown={this.resizeRight}
        />
        <div
          className="brainhub-image-editor__resize-point brainhub-image-editor__resize-point--br"
          onMouseDown={this.resizeBottomRight}
        />
        <div
          className="brainhub-image-editor__resize-point brainhub-image-editor__resize-point--b"
          onMouseDown={this.resizeBottom}
        />
        <div
          className="brainhub-image-editor__resize-point brainhub-image-editor__resize-point--bl"
          onMouseDown={this.resizeBottomLeft}
        />
        <div
          className="brainhub-image-editor__resize-point brainhub-image-editor__resize-point--l"
          onMouseDown={this.resizeLeft}
        />
      </div>
    );
  }
}

SelectedArea.defaultProps = {
  style: {
    top: '0',
    right: '0',
    bottom: '0',
    left: '0',
  },
};

SelectedArea.propTypes = {
  style: PropTypes.shape({
    top: PropTypes.string,
    right: PropTypes.string,
    bottom: PropTypes.string,
    left: PropTypes.string,
  }),
  startMove: PropTypes.func.isRequired,
  startResize: PropTypes.func.isRequired,
};

export default SelectedArea;
