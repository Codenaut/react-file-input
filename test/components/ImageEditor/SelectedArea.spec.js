import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';

import SelectedArea from '../../../src/components/ImageEditor/SelectedArea';

const defaultProps = {
  startMove: () => null,
  startResize: () => null,
};

const setup = (props = {}) => {
  const selectedArea = mount(<SelectedArea {...defaultProps} {...props} />);

  return {
    selectedArea,
    points: selectedArea.find('div.brainhub-selected-area__resize-point'),
  };
};

describe('components', () => {
  describe('ImageEditor', () => {
    describe('SelectedArea', () => {
      it('should render all the resize points', () => {
        const { points } = setup();

        expect(points).toHaveLength(8);
      });

      it('should render with the reight size', () => {
        const style = { top: '10%', bottom: '50%', right: '20%', left: '30%' };
        const { selectedArea } = setup({ style });

        expect(selectedArea.find('div').first().prop('style')).toEqual(style);
      });

      it('should start moving if user clicks on the selectedArea', () => {
        const startMove = jest.fn();
        const { selectedArea } = setup({ startMove });
        selectedArea.simulate('mousedown', { target: selectedArea.find('div').first().instance() });

        expect(startMove).toHaveBeenCalled();
      });

      it('should start resizing if user clicks on the points', () => {
        const startMove = jest.fn();
        const startResize = jest.fn();
        const { points } = setup({ startMove, startResize });

        points.first().simulate('mousedown');

        expect(startMove).not.toHaveBeenCalled();
        expect(startResize).toHaveBeenCalled();
      });

      it('should match exact snapshot', () => {
        const tree = renderer.create(
          <div>
            <SelectedArea {...defaultProps} />
          </div>
        ).toJSON();

        expect(tree).toMatchSnapshot();
      });
    });
  });
});
