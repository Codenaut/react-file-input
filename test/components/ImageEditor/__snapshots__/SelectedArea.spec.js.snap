// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components ImageEditor SelectedArea should match exact snapshot 1`] = `
<div>
  <div
    className="brainhub-selected-area"
    onMouseDown={[Function]}
    style={
      Object {
        "bottom": "0",
        "left": "0",
        "right": "0",
        "top": "0",
      }
    }
  >
    <div
      className="brainhub-selected-area__resize-point brainhub-selected-area__resize-point--tl"
      onMouseDown={[Function]}
    />
    <div
      className="brainhub-selected-area__resize-point brainhub-selected-area__resize-point--t"
      onMouseDown={[Function]}
    />
    <div
      className="brainhub-selected-area__resize-point brainhub-selected-area__resize-point--tr"
      onMouseDown={[Function]}
    />
    <div
      className="brainhub-selected-area__resize-point brainhub-selected-area__resize-point--r"
      onMouseDown={[Function]}
    />
    <div
      className="brainhub-selected-area__resize-point brainhub-selected-area__resize-point--br"
      onMouseDown={[Function]}
    />
    <div
      className="brainhub-selected-area__resize-point brainhub-selected-area__resize-point--b"
      onMouseDown={[Function]}
    />
    <div
      className="brainhub-selected-area__resize-point brainhub-selected-area__resize-point--bl"
      onMouseDown={[Function]}
    />
    <div
      className="brainhub-selected-area__resize-point brainhub-selected-area__resize-point--l"
      onMouseDown={[Function]}
    />
  </div>
</div>
`;
